<?xml version="1.0"?>
<launch>
  <arg name="GROUND_STATION_IP" default="$(env ROS_IP)" />
  <arg name="UAV_IP" default="192.168.1.110" />
  <arg name="HIGH_PORT1" default="16484" />
  <arg name="HIGH_PORT2" default="8192" />
  <arg name="LOW_PORT" default="1024" />
  <arg name="model_suffix" default="_ring_gripper"/>
  <arg name="robot_ns" default="hydrus"/>

  <group ns="$(arg robot_ns)">
    ### UAV to GroundStation
    <node pkg="jsk_network_tools" type="silverhammer_highspeed_receiver.py" name="receiver">
      <param name="message" value="mbzirc2020_task2_common/FC2OCSData" />
      <param name="receive_ip" value="$(arg GROUND_STATION_IP)" />
      <param name="receive_port" value="$(arg HIGH_PORT1)" />
      <param name="topic_prefix" value="" />
    </node>

    ### GroundStation to UAV
    <node pkg="jsk_network_tools" type="silverhammer_lowspeed_streamer.py" name="lowspeed_streamer">
      <remap from="~input" to="/task_start" />
      <param name="message" value="std_msgs/Empty" />
      <param name="to_ip" value="$(arg UAV_IP)" />
      <param name="to_port" value="$(arg LOW_PORT)" />
      <param name="send_rate" value="1" />
      <param name="event_driven" value="True" />
    </node>

    ### GroundStation to UAV
    <node pkg="jsk_network_tools" type="silverhammer_highspeed_streamer.py" name="highspeed_streamer">
      <param name="message" value="mbzirc2020_task2_common/OCS2FCData" />
      <param name="to_ip" value="$(arg UAV_IP)" />
      <param name="to_port" value="$(arg HIGH_PORT2)" />
      <param name="send_rate" value="20" />
      <param name="bandwidth" value="200000" />
      <param name="packet_size" value="1000" />
    </node>

    <node name="image_republish1" pkg="image_transport" type="republish" args="compressed raw">
      <remap from="in" to="rs_d435/color/image_rect_color/throttle" />
      <remap from="out" to="rs_d435/color/image_rect_color/throttle" />
    </node>

    <node name="image_republish2" pkg="image_transport" type="republish" args="compressedDepth raw">
      <remap from="in" to="rs_d435/aligned_depth_to_color/image_raw/throttle" />
      <remap from="out" to="rs_d435/aligned_depth_to_color/image_raw/throttle" />
    </node>

    <node pkg="mbzirc2020_task2_common" type="ground_station_tf_publisher.py" name="ground_station_tf_publisher">
      <param name="robot_ns" value="$(arg robot_ns)"/>
    </node>

    <node pkg="mbzirc2020_task2_common" type="yaml_to_camera_info_publisher.py" name="camera_info_publisher">
      <param name="camera_info_file" value="$(find mbzirc2020_task2_common)/config/camera/RSD435ColorCameraInfo.yaml"/>
      <remap from="~camera_info" to="rs_d435/color/camera_info"/>
    </node>

    <node pkg="nodelet" type="nodelet" name="cloudify" output="screen"
          args="standalone depth_image_proc/point_cloud_xyzrgb" >
      <remap from="rgb/camera_info" to="rs_d435/color/camera_info"/>
      <remap from="rgb/image_rect_color" to="rs_d435/color/image_rect_color/throttle"/>
      <remap from="depth_registered/image_rect" to="rs_d435/aligned_depth_to_color/image_raw/throttle"/>
      <remap from="depth_registered/points" to="rs_d435/points"/>
      <param name="queue_size" value="100"/>
    </node>

    <node pkg="joy" type="joy_node" name="joy_node"  output="screen" >
      <param name="dev" type="string" value="/dev/input/js0"/>
    </node>

    <include file="$(find mbzirc2020_task2_common)/launch/additional/gps_tf_publisher.launch"/>
  </group>

  ###########  Robot Model  ###########
  <include file="$(find aerial_robot_model)/launch/aerial_robot_model.launch">
    <arg name="headless" value="false"/>
    <arg name="robot_model" value="$(find mbzirc2020_task2_common)/robots/hydrus$(arg model_suffix).urdf.xacro"/>
    <arg name="robot_ns" value="$(arg robot_ns)" />
    <arg name="rviz_config" value="$(find mbzirc2020_task2_common)/launch/communication/ground_station_rviz_config"/>
    <arg name="rviz_init_pose" value="$(find mbzirc2020_common)/config/RvizInit.yaml"/>
    <arg name="need_joint_state" value="false"/>
  </include>

</launch>
